### ISSUE TITLE
byte-order problem with X11 backend on big-endian host

### ISSUE THREAD

**Issue by nbriggs** [libsdl-org/SDL#5081]
Drawing with the following environment --
```
       sdl_windowsurface = SDL_GetWindowSurface(sdl_window);
       sdl_pixelformat = sdl_windowsurface->format;
       sdl_black = SDL_MapRGB(sdl_pixelformat, 0, 0, 0);
       sdl_white = SDL_MapRGB(sdl_pixelformat, 255, 255, 255);
       sdl_foreground = sdl_black;
       sdl_background = sdl_white;
[...]
       Uint32 *dst = (Uint32 *)sdl_windowsurface->pixels;
[...]
       dst[dy + dx + b] = (srcw & bitmask[b]) ? sdl_foreground : sdl_background;

```
where
```
(gdb) print sdl_black
$1 = 0
(gdb) print sdl_white
$2 = 16777215
(gdb) print/x sdl_white
$3 = 0xffffff
(gdb) print *sdl_pixelformat
$4 = {format = 370546692, palette = 0x0, BitsPerPixel = 32 ' ', BytesPerPixel = 4 '\004', padding = "\000", Rmask = 16711680, Gmask = 65280, Bmask = 255, Amask = 0, Rloss = 0 '\000', 
  Gloss = 0 '\000', Bloss = 0 '\000', Aloss = 8 '\b', Rshift = 16 '\020', Gshift = 8 '\b', Bshift = 0 '\000', Ashift = 0 '\000', refcount = 2, next = 0x0}
(gdb) print SDL_GetPixelFormatName(370546692)
$5 = 0xffffffff5b274a28 "SDL_PIXELFORMAT_RGB888"
```
this code is running on a big-endian system (Solaris 2.11 on SPARC) but the display is back via an ssh connection from a (little-endian) mac.

xtruss shows that it is sending:
```
00c00000: PutImage(drawable=wp#00C00002, gc=g#00C00003, depth=24, width=1023, height=64, dst-x=0, dst-y=0, left-pad=0, format=ZPixmap, image-data=00000000:FFFFFF00:FFFFFF00:FFFFFF00:00000000:FFFFFF00:FFFFFF00:FFFFFF00:00000000:FFFFFF00:FFFFFF00:FFFFFF00:00000000:...)
```
and sure enough, the display comes up in black (0x000000) and yellow (0xFFFF00) instead of black and white.
If instead of using `SDL_MapRGB(sdl_pixelformat, 255, 255, 255)` for white I use `0xFFFFFFFF` I get the correct black and white display.  I'm not sure which part of SDL is confused.

**Comment by sezero** [comment]
Does https://github.com/libsdl-org/SDL/pull/4867 of @sulix have anything to do with this by any chance?

**Comment by nbriggs** [comment]
I'm not *sure* one way or the other.  It's 4-byte pixels, the black/white should have been in the right format for the pixels in the window surface, but I suspect they got translated incorrectly to the X11 ZPixmap format.

**Comment by slouken** [comment]
I think right now we assume that the host and client endianness are the same. The window surface format should reflect the server's endianness (i.e. flipped if they differ), but I'm not sure where we can query that.

**Comment by nbriggs** [comment]
Looking at the code, it gets the pixel depth and format for the X11 visual it's working with:
```
70          /* Find out the pixel format and depth */
71          if (X11_GetVisualInfoFromVisual(display, data->visual, &vinfo) < 0) {
72              return SDL_SetError("Couldn't get window visual information");
73          }
74      
75          *format = X11_GetPixelFormatFromVisualInfo(display, &vinfo);
```
and in this case
```
(gdb) print *format
$3 = 370546692
(gdb) print vinfo
$4 = {visual = 0x1001b8560, visualid = 34, screen = 0, depth = 24, class = 4, red_mask = 16711680, green_mask = 65280, blue_mask = 255, colormap_size = 256, bits_per_rgb = 8}
(gdb) print SDL_GetPixelFormatName(370546692)
$5 = 0xffffffff5b274a28 "SDL_PIXELFORMAT_RGB888"
```
which is consistent with the black and white values I got above.  Thinking about it, I don't think that the client should ever need to know the endianness of the server, given it should be operating with the visual info it got as the interpretation of the pixels.  What I find odd is that it's attempting (and succeeding) in using the MIT-SHM extension -- which makes no sense to me, since the X server it's talking to is on the other end of an SSH connection...

**Comment by slouken** [comment]
But the mask is in terms of native word order, so the in-memory layout of 0xFF0000 is different between a little and big endian system.

**Comment by nbriggs** [comment]
If the values had been stored in native byte order into the XImage 32-bit pixels then it would have done the right thing (e.g., putting 0xFF.FF.FF in memory as 00.FF.FF.FF -- since this is a big-endian machine) but it put them in memory as FF.FF.FF.00 which is wrong, and then the X11 on the server side interpreted that as xx.RR.GG.BB and so came up with yellow (xx.FF.FF.00).

**Comment by sulix** [comment]
You could try this (totally, 100% untested) patch:
https://github.com/sulix/SDL/commit/bf2d21219af682053f19c8802d6d6fcd3f8d7f16

Apparently just setting the XImage's byte_order field will make the X server do endianness conversion for you:
- https://xorg.freedesktop.narkive.com/GbSD1aPq/ximage-s-byte-order-field
- https://comp.windows.x.narkive.com/aJgKuJw6/endianness-and-bits-per-pixel-abstraction

**Comment by barracuda156** [comment]
I am getting a very similar issue on a Big-endian Mac with `mpv` plus X11. Weirdly, if I use SDL via X11, colors are correct. If I use X11 directly, I get all in yellow.
